# Nome: Antonio Cruciani                Matricola : 0258412 

# Esercizio 3 

In [2]:
import copy 
import itertools
import time
import sys
import matplotlib.pyplot as plt

In [3]:
print sys.version

2.7.12 |Anaconda 4.1.1 (x86_64)| (default, Jul  2 2016, 17:43:17) 
[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)]


## Versione di python utilizzata 
2.7.12 |Anaconda 4.1.1 (x86_64)| (default, Jul  2 2016, 17:43:17) 
[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)]

## Definizione della struttura dati Grafo e delle funzioni per leggere e scrivere su file

In [4]:
#Struttura dati per grafo (Lista di adiacenza)
class Graph:

    def __init__(self, vertices, is_undirected=True):
        self.__v = vertices  # numero di vertici
        self.__edge_list = []  # salva gli archi e gli eventuali pesi (non serve)
        self.__is_undirected = is_undirected # Valore = True per grafi non diretti
        #self.__adj_matrix = None 
        self.__adj_list = None # Lista di adiacenza
        self.__vertici = []#Insieme dei vertici che sono collegati da nodi (non disconnessi)
        self.__covers = [] #Insieme dei nodi coperti 
        
    # Metodo che permette di aggiungere un nuvo aro al grafo
    def add_edge(self, u, v, w=None):
        self.__edge_list.append([u, v, w if w else 1])
        # Se G non è un Digrafo, 
        # copia gli archi nella direzione opposta
        if self.__is_undirected:
            self.__edge_list.append([v, u, w if w else 1])

    # Metodo che costruisce la lista di adiacenza
    def make_adjacency_list(self):
        adj_list = {key: [] for key in range(self.__v)}
        for edge in self.__edge_list:
            # Dove edge[1] è  la destinazione ed un eventuale edge[2] il peso
            edge_val = {edge[1]} 
            adj_list[edge[0]].append(edge_val)
        self.__adj_list = adj_list
        vertici = list()
        for keyvalue in self.__adj_list.items():
            if len(keyvalue[1])!=0: 
                vertici.append(keyvalue[0])
        self.__vertici = vertici
        
    # Metodo che ritorna la lista dei vertici
    def get_vertex_list(self):
        vertici = list()
        
        adj_list = self.get_adj_list()
        for key in range(self.__v):
            if len(adj_list[key])!=0: 
                vertici.append(key)
               
        self.__vertici = vertici
        
    #Metodo che inizializza il cover di tutti i nodi a falso
    def set_false_cover_list(self):
        vertici = list()
        adj_list = self.get_adj_list()
        for key in range(self.__v):

            if(len(adj_list[key])!=0):
                vertici.append([key])
                
        for i in range(0,len(vertici)):
            vertici[i].append("False")
        self.__covers=vertici
        #print vertici
    # Metodo che ritorna la lista dei nodi coperti    
    def get_cover_list(self):
        return self.__covers
    
    #Metodo che marca il nodo scelto e marca gli archi
    def cover(self,lista_nodi):
        #print self.__covers
        for i in lista_nodi:
            self.__covers[i]="True"
            
        return self.__covers
    
    #Il seguente metodo ritorna la lista di adiacenza 
    def get_adj_list(self):
        return self.__adj_list
    
    #Il seguente metodo permette di eliminare un nodo dalla lista di adiacenza
    def remove_edge(self,u):
        del self.__adj_list[u]
        adj={key:[]for key in range(self.__v)}
        #print adj
        for k, v in self.__adj_list.items():
            for h in v:
                if not(u in h):
                    #print "chiave=",k,"valore=", v
                    #print h
                    adj[k].append(h)
                    #print adj[k]
        #print " stampo adj"
        #print adj
        #print type(adj)
        self.__adj_list = adj
        #print self.__adj_list
        
    # Metodo che ritorna la lunghezza della lista definita sul nodo u
    # Ovvero metodo che permette di calcolare il grado del nodo u
    def get_lenvalues(self,u):
        return len(self.__adj_list[u])
    def get_vertex_list(self):
        return self.__vertici
    
    # Metodo che ritorna la chiave della lista di adiacenza che ha il maggior numero di valori nella lista
    # Ovvero, il nodo con grado maggiore.
    def get_key_max_degree(self):
        l = 0
        k = 0
        for keyvalue in self.__adj_list.items():
            key, value = keyvalue[0], keyvalue[1]
            #print key,value,len(value)
            if len(value) >= l:
                l = len(value)
                k = key
        return k
        #return self.__adj_list[max(len(self.__adj_list))]

    # Metodo che ritorna il grado del nodo di grado massimo
    def get_value_max_degree(self):
        l = 0
        k = 0
        for keyvalue in self.__adj_list.items():
            value = keyvalue[1]
            #print key,value,len(value)
            if len(value) >= l:
                l = len(value)
        return l
        
    #Metodo che ritorna due nodi qualsiasi connessi tra di loro
    def get_node(self):
        for keyvalue in self.__adj_list.items():
            if len(keyvalue[1])!=0:
                #print keyvalue[1]
                return keyvalue[0],keyvalue[1][0]
    
#Funzione che legge i parametri di input dal file input.txt e costruisce l'istanza G=(V,E)      
def get_input_param(path = None):
    vertices=list()
    if(path == None):
        source = 'input.txt'
    else:
        source = path
        #print path
    with open(source,'r') as f:
        #j=0
        i=0
        for line in f:
            #print line
            k=list()
            j=0
            for word in line.split():
                if j == 0 and i ==0:
                    n = int(word)
                    G = Graph(n)
                elif j == 1 and i == 0:
                    m = int(word)
                else: 
                    if word != line.split():
                        k.append(int(word)) 

                j+=1
            if i>0:
                if len(k)>0:
                    G.add_edge(int(k[0]),int(k[1]))
            i+=1
        G.make_adjacency_list()
        G.get_vertex_list()
        G.set_false_cover_list()
        return G
# Funzione che scrive i risultati sul file output.txt o su un file output riguardante una classe di grafo random  
def write_on_file_results(solution,Greedy = False, Matching = False, Bforce = False,path= None,d=None):
    if(Greedy == False and Matching == False and Bforce == False and path== False and d==None):
        return -1
    if(Greedy == True and Matching == False and Bforce == False and path== None and d==None):
        outF = open("OutputGreedy.txt", "w")
    if(Matching == True and Greedy == False and Bforce == False and path== None and d==None):
        outF = open("OutputMatching.txt", "w")
    if(Matching == False and Greedy == False and Bforce == True and path== None and d==None):
        outF = open("OutputBruteForce.txt", "w")
    if(Greedy == False and Matching == False and Bforce == False and path != None and d!=None):
        return -1
    if(Greedy == True and Matching == False and Bforce == False and path != None and d!=None):
        outF = open(path+"OutputGreedy_"+d, "w")
    if(Matching == True and Greedy == False and Bforce == False and path != None and d!=None):
        outF = open(path+"OutputMatching_"+d, "w")
    if(Matching == False and Greedy == False and Bforce == True and path != None and d!=None):
        outF = open(path+"OutputBruteForce_"+d, "w")
    s = map(str, solution) 
    k = [str(len(solution))]
    
    textList =k+s
    #print textList
    for line in textList:
        outF.write(line)
        outF.write("\n")
    outF.close()    
   


 # Funzione che ritorna l'inseme delle parti di una lista
def get_power_set(s):
    power_set=[[]]
    for elem in s:
    # iterate over the sub sets so far
        for sub_set in power_set:
              # add a new subset consisting of the subset at hand added elem
            power_set=power_set+[list(sub_set)+[elem]]
    power_set.pop(0)
    return power_set

# print sorted solutions
def get_sorted_solutions(lista):
    h = sorted(lista, key=lambda x: x[1])
    for i in range(0,4):
        print li[i][0:2]
# Funzione che stampa le soluzioni dei due algoritmi: greedy e matching
def print_solutions(lista_1,lista_2):
    n =['50','100','500','1000']

    print "+n----------Greedy------------Matching-------------+"
    for i in range(0,4):
            print n[i],"----------",lista_1[i][1],"------------",lista_2[i][1],"-------------|"
    print "+---------------------------------------------------+"

## Algortimo Greedy per la risoluzione del Vertex Cover

In [5]:
def GreedyMinVertexCover(source = None):
    G= get_input_param(source)
    G.get_adj_list()
    S = list()
    start = time.time()
    while len(G.get_adj_list())!=0:
            k = G.get_key_max_degree()
            #print k
            #print G.get_adj_list()
            if G.get_lenvalues(k) != 0:
                S.append(k)
            else:
                break
            G.remove_edge(k)
            #print G.get_adj_list()
    stop = time.time()
    duration = stop-start
    print "Il min Vertex Cover è =",S,"La sua cardinalità è =",len(S),"L'algoritmo Greedy Min Vertex Cover ha impiegato = ",duration
    return S,duration

## Algoritmo di Matching per Vertex Cover

In [6]:
def Matching(source = None):
    
    G = get_input_param(source)
    
    S =list()
    
    start=time.time()
    while G.get_lenvalues(G.get_key_max_degree()) !=0:
        #print G.get_adj_list()
        u,v1 = G.get_node()
        
        v = list(v1)
        p = int(v[0])

        G.remove_edge(u)
        G.remove_edge(p)
        S.append(u)
        S.append(p)
    stop = time.time()
    duration = stop-start
    print "Il min Vertex Cover è =",S," La sua cardinalità è =",len(S),"L'algoritmo Matching Min Vertex Cover ha impiegato = ",duration
    return S,duration
        
        

## Algoritmo Esaustivo per Vertex Cover

In [7]:
def MinVertexCoverExhaustive(source = None):
    G = get_input_param(source)
    #Generazione dell'insieme delle parti 
    ps = get_power_set(G.get_vertex_list())
    j=0
    k= float("inf")
    start = time.time()
    for i in ps:
        grafo=get_input_param(source)
        #print grafo.get_adj_list()
        #i.remove_edge(u)
        #print ps[j]
        for c in ps[j]:
            grafo.remove_edge(c)
            
        grafo.cover(ps[j])
        
        if grafo.get_value_max_degree() == 0:
            
            if k>len(ps[j]):
                k=len(ps[j])
                index = j
        j+=1
    stop = time.time()
    duration = stop-start
    print "Il min Vertex Cover è =",ps[index],"La sua cardinalità è =",k," Il tempo impiegato dall'algoritmo esaustivo è =",duration
    return ps[index],duration
    
   
    

## Esecuzione dei tre Algoritmi

In [9]:
sol_gmvc,dur_g = GreedyMinVertexCover()
match,dur_m = Matching()


Il min Vertex Cover è = [0, 4, 3] La sua cardinalità è = 3 L'algoritmo Greedy Min Vertex Cover ha impiegato =  0.000123977661133
Il min Vertex Cover è = [0, 1, 2, 4]  La sua cardinalità è = 4 L'algoritmo Matching Min Vertex Cover ha impiegato =  5.79357147217e-05


In [120]:
# Esecuzione dell' algoritmo esaustivo
BF,dur_b = MinVertexCoverExhaustive()

Il min Vertex Cover è = [0, 2, 3, 4] La sua cardinalità è = 4  Il tempo impiegato dall'algoritmo esaustivo è = 0.0719020366669


## Scrittura dei risultati su file
### Greedy -> outuputGreedy.txt
### Matching -> outputMatching.txt
### Brute Force -> outputBruteForce.txt

In [11]:
write_on_file_results(sol_gmvc,Greedy = True)
write_on_file_results(match,Matching = True)
write_on_file_results(BF,Bforce = True)

# Comparazione performance dei tre algoritmi su input.txt

In [122]:
l=[["Durata Greedy",dur_g],["Durata Matching",dur_m],["Durata Bruteforce",dur_b]]

In [123]:
sorted(l, key=lambda x: x[1])

[['Durata Matching', 5.793571472167969e-05],
 ['Durata Bruteforce', 0.07190203666687012],
 ['Durata Greedy', 1.3217549324035645]]

## Test Bruteforce con istanza dandom con P=$\frac{1}{2}$ ed n=20

In [10]:
BF,dur_b = MinVertexCoverExhaustive("RandomGraphs/un/bf/input_20.txt")

KeyboardInterrupt: 

L'algoritmo impiega più di 10 secondi

# Test delle performance su grafi random

Verranno creati dei files di output situati nella cartella Outputs

## Esecuzione su grafi con n = 50, 100, 500 ,1000 e p = $\frac{1}{2}$

In [43]:
n=[50,100,200,500,1000]

In [98]:
#Simulazione Greedy per i diversi valori di n
path ='RandomGraphs/un/'
lista_greedy_un =[]
fil = ['input_50.txt','input_100.txt','input_500.txt','input_1000.txt']
for i in fil:
    dest = path+i
    #print dest
    out_dest = 'Outputs/Greedy/un/'
    sol_gmvc,dur_g = GreedyMinVertexCover(dest)
    lista_greedy_un.append([i,dur_g,sol_gmvc])
    write_on_file_results(sol_gmvc,Greedy = True,path=out_dest,d=i)
#Simulazione Matching per i diversi valori di n
lista_match_un =[]
path ='RandomGraphs/un/'
for i in fil:
    dest = path+i
    #print dest
    out_dest = 'Outputs/Matching/un/'
    sol_mat,dur_g = Matching(dest)
    lista_match_un.append([i,dur_g,sol_mat])
    write_on_file_results(sol_mat,Matching = True,path=out_dest,d=i)

Il min Vertex Cover è = [1, 23, 41, 31, 37, 25, 10, 32, 16, 30, 4, 22, 18, 49, 43, 42, 40, 44, 19, 35, 12, 9, 27, 11, 8, 34, 13, 45, 21, 33, 7, 48, 39, 0, 15, 14, 38, 28, 2, 46, 20, 26, 6, 5, 17, 47, 36, 29] La sua cardinalità è = 48 L'algoritmo Greedy Min Vertex Cover ha impiegato =  0.0522010326385
Il min Vertex Cover è = [73, 24, 46, 9, 7, 85, 61, 4, 76, 33, 10, 45, 75, 91, 56, 29, 74, 53, 96, 31, 20, 12, 70, 57, 86, 71, 50, 40, 17, 3, 68, 18, 88, 22, 90, 26, 89, 84, 16, 37, 78, 30, 14, 34, 80, 13, 38, 32, 95, 43, 47, 2, 15, 51, 94, 59, 44, 64, 60, 87, 25, 81, 66, 62, 27, 23, 39, 35, 54, 48, 55, 52, 6, 0, 67, 72, 63, 36, 99, 19, 5, 28, 21, 8, 97, 77, 82, 1, 11, 92, 93, 69, 65, 42, 49, 98, 41, 83] La sua cardinalità è = 98 L'algoritmo Greedy Min Vertex Cover ha impiegato =  0.479500055313
Il min Vertex Cover è = [2, 90, 17, 29, 9, 66, 6, 26, 105, 35, 69, 54, 111, 79, 1, 89, 44, 31, 113, 81, 13, 27, 115, 52, 3, 112, 53, 10, 51, 97, 62, 41, 92, 84, 55, 99, 102, 85, 78, 24, 45, 38, 101,

In [99]:
print_solutions(lista_greedy_un,lista_match_un)

+n----------Greedy------------Matching-------------+
50 ---------- 0.0522010326385 ------------ 0.0805299282074 -------------|
100 ---------- 0.479500055313 ------------ 0.485810995102 -------------|
500 ---------- 54.3053991795 ------------ 53.9215009212 -------------|
1000 ---------- 134.326865911 ------------ 125.709211826 -------------|
+---------------------------------------------------+


## Esecuzione su grafi con n = 50, 100, 200, 500, 1000 con $p = \frac{1}{\sqrt{n}}$

In [100]:
#Simulazione Greedy per i diversi valori di n
path ='RandomGraphs/sqrt/'
lista_greedy_sqrt =[]
fil = ['input_50.txt','input_100.txt','input_500.txt','input_1000.txt']
for i in fil:
    dest = path+i
    #print dest
    out_dest = 'Outputs/Greedy/sqrt/'
    sol_gmvc,dur_g = GreedyMinVertexCover(dest)
    lista_greedy_sqrt.append([i,dur_g,sol_gmvc])
    write_on_file_results(sol_gmvc,Greedy = True,path=out_dest,d=i)
#Simulazione Matching per i diversi valori di n
lista_match_sqrt =[]
path ='RandomGraphs/sqrt/'
for i in fil:
    dest = path+i
    #print dest
    out_dest = 'Outputs/Matching/sqrt/'
    sol_mat,dur_g = Matching(dest)
    lista_match_sqrt.append([i,dur_g,sol_mat])
    write_on_file_results(sol_mat,Matching = True,path=out_dest,d=i)

Il min Vertex Cover è = [26, 10, 31, 16, 5, 11, 45, 42, 0, 7, 13, 38, 18, 6, 17, 48, 25, 15, 32, 9, 4, 47, 46, 29, 23, 21, 43, 35, 36, 28, 8, 34, 19, 22, 20, 14, 44, 30, 24] La sua cardinalità è = 39 L'algoritmo Greedy Min Vertex Cover ha impiegato =  0.0155920982361
Il min Vertex Cover è = [92, 70, 42, 87, 52, 88, 81, 35, 46, 94, 93, 84, 33, 31, 74, 82, 11, 63, 41, 30, 49, 48, 1, 73, 62, 24, 4, 83, 8, 91, 5, 0, 77, 97, 71, 36, 80, 78, 55, 25, 6, 23, 18, 14, 44, 34, 21, 12, 85, 66, 19, 7, 65, 61, 56, 45, 13, 90, 67, 32, 59, 29, 28, 9, 98, 68, 58, 57, 50, 10, 3, 53, 15, 2, 89, 69, 43, 17, 79, 76, 64, 60, 27] La sua cardinalità è = 83 L'algoritmo Greedy Min Vertex Cover ha impiegato =  0.0436961650848
Il min Vertex Cover è = [489, 462, 55, 341, 265, 200, 125, 464, 198, 114, 314, 261, 204, 79, 368, 66, 465, 449, 307, 260, 226, 48, 435, 425, 274, 165, 61, 334, 224, 431, 418, 318, 86, 493, 291, 193, 135, 103, 37, 451, 448, 399, 347, 322, 159, 476, 319, 303, 129, 100, 0, 458, 402, 230, 182, 

In [101]:
print_solutions(lista_greedy_sqrt,lista_match_sqrt)

+n----------Greedy------------Matching-------------+
50 ---------- 0.0155920982361 ------------ 0.0091769695282 -------------|
100 ---------- 0.0436961650848 ------------ 0.0482280254364 -------------|
500 ---------- 2.75884008408 ------------ 2.86245799065 -------------|
1000 ---------- 14.9498388767 ------------ 17.0957140923 -------------|
+---------------------------------------------------+


## Esecuzione su grafi con n = 50, 100, 200, 500, 1000 con $p = \frac{log(n)}{n}$

In [102]:
#Simulazione Greedy per i diversi valori di n
path ='RandomGraphs/log/'
lista_greedy_log =[]
fil = ['input_50.txt','input_100.txt','input_500.txt','input_1000.txt']
for i in fil:
    dest = path+i
    #print dest
    out_dest = 'Outputs/Greedy/log/'
    sol_gmvc,dur_g = GreedyMinVertexCover(dest)
    lista_greedy_log.append([i,dur_g,sol_gmvc])
    write_on_file_results(sol_gmvc,Greedy = True,path=out_dest,d=i)
#Simulazione Matching per i diversi valori di n
lista_match_log =[]
path ='RandomGraphs/log/'
for i in fil:
    dest = path+i
    #print dest
    out_dest = 'Outputs/Matching/log/'
    sol_mat,dur_g = Matching(dest)
    lista_match_log.append([i,dur_g,sol_mat])
    write_on_file_results(sol_mat,Matching = True,path=out_dest,d=i)

Il min Vertex Cover è = [38, 20, 32, 46, 23, 17, 21, 31, 30, 22, 13, 1, 8, 0, 19, 4, 48, 39, 24, 47, 34, 16, 9, 7, 6, 42, 18, 14, 12, 45, 11, 2, 49, 44, 36, 25] La sua cardinalità è = 36 L'algoritmo Greedy Min Vertex Cover ha impiegato =  0.00817203521729
Il min Vertex Cover è = [25, 74, 68, 23, 89, 79, 11, 55, 22, 71, 26, 49, 28, 13, 80, 37, 1, 94, 90, 14, 3, 87, 73, 44, 16, 92, 78, 69, 7, 0, 61, 58, 46, 32, 31, 21, 99, 60, 59, 47, 41, 19, 95, 9, 96, 93, 85, 76, 70, 64, 63, 57, 54, 53, 24, 10, 88, 86, 51, 38, 35, 29, 20, 8, 83, 77, 67, 50, 15] La sua cardinalità è = 69 L'algoritmo Greedy Min Vertex Cover ha impiegato =  0.0312669277191
Il min Vertex Cover è = [361, 174, 425, 217, 118, 491, 458, 226, 96, 6, 498, 358, 430, 323, 220, 213, 200, 113, 110, 480, 464, 387, 365, 332, 322, 317, 223, 204, 153, 137, 135, 131, 469, 368, 303, 229, 221, 189, 161, 460, 400, 341, 314, 252, 243, 228, 203, 168, 162, 149, 71, 69, 28, 499, 495, 486, 459, 448, 439, 435, 408, 373, 370, 330, 242, 225, 58, 37

In [103]:
print_solutions(lista_greedy_log,lista_match_log)

+n----------Greedy------------Matching-------------+
50 ---------- 0.00817203521729 ------------ 0.00884294509888 -------------|
100 ---------- 0.0312669277191 ------------ 0.0223581790924 -------------|
500 ---------- 0.665121078491 ------------ 0.788973093033 -------------|
1000 ---------- 3.82676005363 ------------ 4.72154712677 -------------|
+---------------------------------------------------+


## Esecuzione su grafi con n = 50, 100, 200, 500, 1000 con $p = \frac{1}{n}$

In [104]:
#Simulazione Greedy per i diversi valori di n
path ='RandomGraphs/inv/'
lista_greedy_inv =[]
fil = ['input_50.txt','input_100.txt','input_500.txt','input_1000.txt']
for i in fil:
    dest = path+i
    #print dest
    out_dest = 'Outputs/Greedy/inv/'
    sol_gmvc,dur_g = GreedyMinVertexCover(dest)
    lista_greedy_inv.append([i,dur_g,sol_gmvc])
    write_on_file_results(sol_gmvc,Greedy = True,path=out_dest,d=i)
#Simulazione Matching per i diversi valori di n
lista_match_inv =[]
path ='RandomGraphs/inv/'
for i in fil:
    dest = path+i
    #print dest
    out_dest = 'Outputs/Matching/inv/'
    sol_mat,dur_g = Matching(dest)
    lista_match_inv.append([i,dur_g,sol_mat])
    write_on_file_results(sol_mat,Matching = True,path=out_dest,d=i)

Il min Vertex Cover è = [42, 19, 47, 46, 41, 8, 3, 45, 43, 40, 38, 37, 17, 49, 39, 35, 33, 32, 29, 23, 20, 18] La sua cardinalità è = 22 L'algoritmo Greedy Min Vertex Cover ha impiegato =  0.00227403640747
Il min Vertex Cover è = [53, 98, 86, 91, 77, 49, 44, 21, 1, 97, 75, 73, 71, 56, 55, 43, 20, 9, 2, 93, 68, 66, 47, 42, 37, 33, 23, 11, 7, 3, 0, 99, 96, 90, 83, 74, 72, 62, 61, 60, 52, 50, 27] La sua cardinalità è = 43 L'algoritmo Greedy Min Vertex Cover ha impiegato =  0.00979495048523
Il min Vertex Cover è = [360, 85, 30, 432, 318, 222, 86, 76, 49, 491, 479, 388, 375, 370, 358, 311, 295, 246, 226, 212, 163, 147, 486, 456, 446, 436, 431, 424, 421, 408, 379, 361, 342, 323, 266, 256, 244, 237, 175, 152, 145, 115, 93, 80, 24, 10, 3, 475, 472, 467, 466, 459, 430, 418, 417, 411, 395, 386, 385, 384, 376, 373, 365, 352, 340, 333, 322, 260, 252, 231, 219, 218, 188, 168, 144, 143, 142, 140, 126, 98, 72, 45, 44, 43, 497, 484, 473, 464, 457, 453, 447, 433, 415, 414, 413, 406, 405, 396, 393, 383,

In [105]:
print_solutions(lista_greedy_inv,lista_match_inv)

+n----------Greedy------------Matching-------------+
50 ---------- 0.00227403640747 ------------ 0.00696897506714 -------------|
100 ---------- 0.00979495048523 ------------ 0.011626958847 -------------|
500 ---------- 0.176158905029 ------------ 0.303361177444 -------------|
1000 ---------- 0.875744104385 ------------ 1.3217549324 -------------|
+---------------------------------------------------+


# Analisi delle perfomance

$p = \frac{1}{n}$
<table style="width:100%">
  <tr>
    <th>n</th>
    <th>Greedy</th> 
    <th>Matching</th>
    <th>Size VC Greedy</th>
    <th>Size VC Matching</th>
  </tr>
  <tr>
    <td>50</td>
    <td>0.0522010326385</td> 
    <td>0.0805299282074</td>
    <td>22</td>
    <td>38</td>
  </tr>
  <tr>
    <td>100</td>
    <td>0.479500055313</td> 
    <td>0.485810995102</td>
    <td>43</td> 
    <td>76</td>
  </tr>
   <tr>
    <td>500</td>
    <td>54.3053991795</td> 
    <td>53.9215009212</td>
    <td>202</td> 
    <td>344</td>
  </tr>
   <tr>
    <td>1000</td>
    <td>134.326865911</td> 
    <td>125.709211826</td>
    <td>405</td> 
    <td>678</td>
  </tr>
</table>

$p = \frac{1}{\sqrt{n}}$

<table style="width:100%">
  <tr>
    <th>n</th>
    <th>Greedy</th> 
    <th>Matching</th>
    <th>Size VC Greedy</th>
    <th>Size VC Matching</th>
  </tr>
  <tr>
    <td>50</td>
    <td>0.0155920982361</td> 
    <td>0.0091769695282</td>
    <td>39</td>
    <td>48</td>
  </tr>
  <tr>
    <td>100</td>
    <td>0.0436961650848</td> 
    <td>0.0482280254364</td>
    <td>83</td> 
    <td>96</td>
  </tr>
   <tr>
    <td>500</td>
    <td>2.75884008408</td> 
    <td>2.86245799065</td>
    <td>452</td> 
    <td>494</td>
  </tr>
   <tr>
    <td>1000</td>
    <td>14.9498388767</td> 
    <td>17.0957140923</td>
    <td>930</td> 
    <td>986</td>
  </tr>
</table>
 
$p = \frac{log(n)}{n}$

<table style="width:100%">
  <tr>
    <th>n</th>
    <th>Greedy</th> 
    <th>Matching</th>
    <th>Size VC Greedy</th>
    <th>Size VC Matching</th>
  </tr>
  <tr>
    <td>50</td>
    <td>0.00817203521729</td> 
    <td>0.00884294509888</td>
    <td>36</td>
    <td>46</td>
  </tr>
  <tr>
    <td>100</td>
    <td>0.0312669277191</td> 
    <td>0.0223581790924</td>
    <td>69</td> 
    <td>94</td>
  </tr>
   <tr>
    <td>500</td>
    <td>0.665121078491</td> 
    <td>0.788973093033</td>
    <td>380</td> 
    <td>472</td>
  </tr>
   <tr>
    <td>1000</td>
    <td>3.82676005363</td> 
    <td>4.72154712677</td>
    <td>764</td> 
    <td>950</td>
  </tr>
</table>

p = $\frac{1}{2}$

<table style="width:100%">
  <tr>
    <th>n</th>
    <th>Greedy</th> 
    <th>Matching</th>
    <th>Size VC Greedy</th>
    <th>Size VC Matching</th>
  </tr>
  <tr>
    <td>50</td>
    <td>0.0155920982361</td> 
    <td>0.0091769695282</td>
    <td>39</td>
    <td>48</td>
  </tr>
  <tr>
    <td>100</td>
    <td>0.0436961650848</td> 
    <td>0.0482280254364</td>
    <td>83</td> 
    <td>96</td>
  </tr>
   <tr>
    <td>500</td>
    <td>2.75884008408</td> 
    <td>2.86245799065</td>
    <td>452</td> 
    <td>494</td>
  </tr>
   <tr>
    <td>1000</td>
    <td>14.9498388767</td> 
    <td>17.0957140923</td>
    <td>930</td> 
    <td>986</td>
  </tr>
</table>
 
$p = \frac{log(n)}{n}$

<table style="width:100%">
  <tr>
    <th>n</th>
    <th>Greedy</th> 
    <th>Matching</th>
    <th>Size VC Greedy</th>
    <th>Size VC Matching</th>
  </tr>
  <tr>
    <td>50</td>
    <td>0.0522010326385</td> 
    <td>0.0805299282074</td>
    <td>48</td>
    <td>50</td>
  </tr>
  <tr>
    <td>100</td>
    <td>0.479500055313</td> 
    <td>0.485810995102</td>
    <td>98</td> 
    <td>100</td>
  </tr>
   <tr>
    <td>500</td>
    <td>54.3053991795</td> 
    <td>53.9215009212</td>
    <td>496</td> 
    <td>500</td>
  </tr>
   <tr>
    <td>1000</td>
    <td>134.326865911</td> 
    <td>125.709211826</td>
    <td>991</td> 
    <td>998</td>
  </tr>
</table>



### Osservazioni
Osserviamo che l'algoritmo greedy trova dei vertex cover di dimensione leggermente più piccola rispetto all'algoritmo di matching.
Per quanto riguada le performance temporali essi differiscono di poco. 

## Osservazioni algoritmo esaustivo
Possiamo utilizzare tale algoritmo per istanze con un numero di nodi non troppo elevato. Ovviamente non è stato possibile effettuare delle simulazioni del bruteforce con la classe di grafi casuali.
Quando arriviamo ad un n relativamente alto, calcolarsi l'insieme delle parti, ovvero, l'insieme di tutti i possibili sottoinsiemi candidati al test di cover, diventa impossibile. 
Ad esempio, utilizzando $n=50$ abbiamo che l'insieme potenza è $2^n$ ed è un insieme di cardinalità 1125899906842624.